cryptography/src/lib.rs

```diff
@@ -5,90 +5,64 @@
 //! `commonware-cryptography` is **ALPHA** software and is not yet recommended for production use. Developers should
 //! expect breaking changes and occasional instability.
 
+use commonware_codec::{Encode, ReadExt};
 use commonware_utils::Array;
 use rand::{CryptoRng, Rng, RngCore, SeedableRng};
 
 pub mod bls12381;
-pub use bls12381::Bls12381;
 pub mod ed25519;
-pub use ed25519::{Ed25519, Ed25519Batch};
 pub mod sha256;
 pub use sha256::{hash, CoreSha256, Sha256};
 pub mod secp256r1;
-pub use secp256r1::Secp256r1;
-
-/// Any implementation of a cryptographic primitive must specify a `PublicKey` and `Signature`.
-pub trait Specification {
-    /// Public key used for verifying signatures.
-    type PublicKey: Array;
-    /// Signature generated by signing a message.
-    type Signature: Array;
-}
 
-/// Implementation that can verify that a `PublicKey` produced a valid `Signature`.
-pub trait Verifier: Specification + Clone + Send + Sync + 'static {
-    /// Check that a signature is valid for the given message and public key.
-    ///
-    /// The message should not be hashed prior to calling this function. If a particular
-    /// scheme requires a payload to be hashed before it is signed, it will be done internally.
-    ///
-    /// Because namespace is prepended to message before signing, the namespace provided here must
-    /// match the namespace provided during signing.
-    fn verify(
-        namespace: Option<&[u8]>,
-        message: &[u8],
-        public_key: &Self::PublicKey,
-        signature: &Self::Signature,
-    ) -> bool;
-}
+/// Produces signatures over messages and has a corresponding public key.
+pub trait Signer: Send + Sync + Clone + 'static {
+    /// The type of signature produced by this signer.
+    type Signature: Signature;
+    /// The corresponding public key type.
+    type PublicKey: PublicKey<Signature = Self::Signature>;
 
-/// Implementation that can sign a message with a `PrivateKey`.
-pub trait Signer: From<Self::PrivateKey> + Specification + Clone + Send + Sync + 'static {
-    /// Private key used for signing.
-    type PrivateKey: Array;
+    /// Returns the public key corresponding to this private key.
+    fn public_key(&self) -> Self::PublicKey;
 
-    /// Returns a new instance of the scheme.
-    fn new<R: Rng + CryptoRng>(rng: &mut R) -> Self;
+    /// Sign a message with the given namespace.
+    fn sign(&self, namespace: Option<&[u8]>, msg: &[u8]) -> Self::Signature;
```

Comment:
I think this should remain `mut` (as some signature schemes can be stateful/require mutation of some state to prevent leaking the secret).

I suppose that could be implemented with an `Arc` tho...

Reply:
I suppose the `Clone` here undermines my argument. Will close.